### PR TITLE
Extract Method Refactoring

### PR DESCRIPTION
**Pull request by tommy9**
This revives and updates the old ExtractMethod branch from @bclothier from 2017. I've tried to update things to match the current style of other refactorings and add remaining functionality. It seems to be working well in my testing but it's pretty complex so no doubt there will be some bugs once others try it out.

There is one major issue currently that I'm not sure how to approach: The dialog box fills the screen vertically, hiding the buttons until the window is resized/moved. No idea what is happening here as it looks okay in the Visual Studio designer.

Most of the logic exists in two places: 

- Rubberduck.Core/Refactorings/ExtractMethod/ExtractMethodModel.cs is where the logic for working out the changes needed to the calling procedure and for building the new procedure. In other refactorings, much of the logic is generally outside of the model (e.g. in the Refactoring file) but I wanted to keep it close to the view model rather than having a chain of properties.
- Rubberduck.Refactorings/ExtractMethod/ExtractMethodSelectionValidation.cs is where initial checks are carried out for code that can't (currently) be refactored e.g. GoTo statements, largely unchanged from the old branch

The logic how to perform the extraction was based on a thesis I found online [(link)](https://ora.ox.ac.uk/objects/uuid:1a027679-1e2b-4fb5-a6ff-3270f15154a1) rather than that described in the [wiki](https://github.com/rubberduck-vba/Rubberduck/wiki/Extract-Method-Refactoring-%3A-Workings---Determining-what-params-to-move). Fundamentally they are very similar of course.

### CODE REVIEW

File: Rubberduck.Refactorings/RefactoringsUI.resx [hunk 1]
```diff
@@ -619,4 +621,86 @@ Do you want to proceed?</value>
...
+    <value>Name</value>
+  </data>
+  <data name="ExtractMethod_UnableToMoveVariableDeclarationMessage" xml:space="preserve">
+    <value>Unable to move variable declaration</value>
+  </data>
+  <data name="ExtractMethod_InvalidMessageSelectionEndsOutsideProcedure" xml:space="preserve">
+    <value>Selection does not end inside a recognised procedure</value>
+  </data>
+  <data name="ExtractMethod_InvalidMessageSelectionHasEndStatement" xml:space="preserve">
+    <value>Extract method cannot extract code that contains an End statement.</value>
```
Comment (retailcoder) [inline Rubberduck.Refactorings/RefactoringsUI.resx:659]: Would probably be easier for translators if it were just one resource key with a format-placeholder, like "Extract Method cannot extract code that contains a '{0}' token."

**Comment by tommy9** [comment]
An update on the changes. The issue with the dialog box filling the screen vertically has been resolved so no major known issues. Specifically on the various issues related to extract method found in the [Current Epic](https://github.com/rubberduck-vba/Rubberduck/projects/2) project, I've tested in Excel with any referenced code with the result...

Fixes #1599 - Get the UI with the ability rename the method, preview changes, extract a function and pick the return value, perform indentation, put empty line before the extracted method and pass unassigned parameters `ByVal`.
Fixes #2283 - Checks the validity of the new method name and adds a numerical suffix if necessary.
Fixes #2202 - New method call in correct place.
Fixes #2284 - Similar to 2203, new method call in correct place.
Fixes #2139 - Will report an error if trying to extract example code since it modifies the property return value instead of creating invalid code.
Fixes #1473 - Can extract code where it is valid to do so.
Fixes #2285 - Indentation is applied to the new method (using call to smart indenter) and new method call (using same indentation as start of extracted code.
Fixes #2326 - Just about re-enabling extract method refactoring.
Fixes #2515 - Will give suitable error message if trying to extract code which contains a `GoTo` statement.
Fixes #2783 - Similar to 2203, new method call in correct place.
Fixes #3509 - Restores refactor menu item.

Limited functionality to deal with #2025. A warning is shown on the dialog if the selection to extract contains any precompiled directives, however it doesn't show a warning if there are any in the module (or in the method to extract from but outside of the selection). Can change the scope of warnings/errors fairly easily if desired.

File: Rubberduck.Refactorings/ExtractMethod/IExtractMethodPresenter.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using Rubberduck.Parsing.Symbols;
+
+namespace Rubberduck.Refactorings.ExtractMethod
+{
+    public interface IExtractMethodPresenter : IRefactoringPresenter<ExtractMethodModel>
+    {
+        //ExtractMethodModel Show(IExtractMethodModel methodModel, IExtractMethodProc extractMethodProc);
```
Comment (retailcoder) [inline Rubberduck.Refactorings/ExtractMethod/IExtractMethodPresenter.cs:7]: Dead code?

File: Rubberduck.VBEEditor/Extensions/CodeModuleExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        // whitespace between it and the start of the line. Labels or line numbers are not legal in those lines.
+
+        /// <summary>
+        /// Indicates whether a given selection within a code module contains conditional compilation directives which
+        /// may impede refactoring of the code
+        /// </summary>
+        /// <param name="codeModule">CodeModule containing the selected code</param>
+        /// <param name="selection">The selected code to test</param>
+        /// <returns>True if any conditional compilation directives are found</returns>
+        public static bool ContainsCompilationDirectives(this ICodeModule codeModule, Selection selection)
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/Extensions/CodeModuleExtensions.cs:21]: Not a showstopper in any way (mostly a note to myself), but for easier porting to the 3.0 code base the calls that acquire the module code should be separate from the string content validation logic.

**Review by retailcoder — APPROVED** [review]
Awesome, pretty much good to go as-is. Thank you so much for this contribution!

**Pull request merged**